docs/debug_tests.rst

```diff
@@ -0,0 +1,81 @@
+.. _topics-debug_tests:
+
+===========
+Debug Tests
+===========
+
+The test results go in ``results/<date>—<test_number>``. For results from a particular test, look for ``results/<date>—<test_number>/test_class_name/<test_method_name>/`` directory. The ``test_log.debug`` file will contain the log output from the test, and logs of services used in the test will be in ``service_name/node_name`` sub-directory.
+
+If there is not enough information in the logs, you can re-run the test with ``—no-teardown`` argument::
+
+    ducktape dir/tests/my_test.py::TestA.test_a —no-teardown
+
+
+This will run the test but will not kill any running processes or remove log files when the test finishes running. Then, you can examine the state of a running service or the machine when the service process is running by logging into that machine. Suppose you suspect a particular service being the cause of the test failure. You can find out which machine was allocated to that service by either looking at ``test_log.debug`` or at directory names under ``results/<date>—<test_number>/test_class_name/<test_method_name>/service_name/``. It could be useful to add an explicit debug log to ``start_node`` method with a node ID and node’s hostname information for easy debugging::
+
+    def start_node(self, node):
+        idx = self.idx(node)
+        self.logger.info("Starting ZK node %d on %s", idx, node.account.hostname)
+
+The log statement will look something like this::
+
+    [INFO  - 2017-03-28 22:07:25,222 - zookeeper - start_node - lineno:50]: Starting ZK node 1 on worker1
+
+If you are using Vagrant for example, you can then log into that node via::
+
+    vagrant ssh worker1
+
+
+
+Use Logging
+===========
+
+Distributed system tests can be difficult to debug. You want to add a lot of logging for debugging and tracking progress of the test. A good approach would be to log an intention of an operation with some useful information before any operation that can fail. Also, put enough information to a message of ``assert`` to help figure out what went wrong. Consider an example of testing ElasticSearch service::
```

Comment:
maybe also mention that it can be a good idea to use a higher logging level than you would in prod so more info is available, e.g. maybe make your log levels default to DEBUG instead of INFO. This is good thing to adjust in the Services themselves rather than at a test level.